torchx/schedulers/ray/ray_driver.py

```diff
@@ -77,77 +111,184 @@ def load_actor_json(filename: str) -> List[RayActor]:
+                self.replicas[replica_ix_of_pg[i] : replica_ix_of_pg[i + 1]]
+            )
+            for i in range(len(replica_ix_of_pg) - 1)
+        ]
+
+    def pop_actor_info(self, actor_id: str) -> ActorInfo:
+        """Remove the info of  (dead) command actor"""
+        return self.actor_info_of_id.pop(actor_id)
 
+    def reschedule_actor(self, actor_id: str) -> None:
+        """Rescheule a failed actor"""
+        # pop the information of failed actor
+        info = self.pop_actor_info(actor_id)
+        _logger.info(f"Rescheduling actor {actor_id} to placement group {info.pg}")
+        self.create_and_schedule_actor(info.pg, info.replica)
 
-def create_command_actors(
-    actors: List[RayActor], pg: PlacementGroup
-) -> List[CommandActor]:
-    cmd_actors: List[CommandActor] = []
-    for i, replica in enumerate(actors):
-        # Environment variables for distributed training
+    def create_and_schedule_actor(self, pg: PlacementGroup, replica: RayActor) -> None:
+        """create an command actor in the given placement group"""
         actor = CommandActor.options(  # pyre-ignore[16]
             placement_group=pg,
             num_cpus=replica.num_cpus,
             num_gpus=replica.num_gpus,
         ).remote(replica.command, replica.env)
-        cmd_actors.append(actor)
-
-        if i == 0:
-            rank_0_address = "localhost"
-            rank_0_port = 0
-        else:
-            rank_0_address, rank_0_port = ray.get(
-                # pyre-ignore[16]
-                cmd_actors[0].get_actor_address_and_port.remote()
-            )
-        ray.get(actor.set_address_and_port.remote(rank_0_address, rank_0_port))
 
-    return cmd_actors
+        actor_id = actor._actor_id.hex()
+        # launch a task to check if the actor is scheduled
+        self.active_tasks.append(actor.schedule.remote(actor_id))
+        self.actor_info_of_id[actor_id] = ActorInfo(
+            actor=actor,
+            pg=pg,
+            replica=replica,
+        )
+
+    def place_command_actors(self) -> None:
+        """Creating pending tasks to initialize command actors in placement groups"""
+        pg_ix_of_replica: List[int] = [
+            max(0, i - self.min_nnodes + 1) for i in range(len(self.replicas))
+        ]  # the placement group indices for corresponding replicas
+        for i in range(len(self.replicas)):
+            pg_ix = pg_ix_of_replica[i]
+            pg = self.placement_groups[pg_ix]  # find the created placement group
+            replica = self.replicas[i]
+            self.create_and_schedule_actor(pg, replica)
+
+    def run(self) -> None:
+        result: Optional[PlacementGroup]  # execution result
+        need_more_actors: bool = True  # if need more actors
+        command_actors_count: int = 0  # number of created command actors
+        # Await return result of remote ray function and initialize new command actors
+        while len(self.active_tasks) > 0:
```

Comment:
can we split loop out from run so it has a "_step()" and "run()" method so we can manually step through the behavior from a mocked unit test?